display/printf.rs

```diff
@@ -63,274 +72,450 @@ enum ParseState {
+        'd' | 'i' => format_arg_int(conv, arg)?,
+        'u' | 'o' | 'x' | 'X' => format_arg_uint_base(conv, arg)?,
         'c' => format_arg_char(conv, arg),
         's' => format_arg_string(conv, arg),
+        ch => {
+            eprintln!("printf: unknown conversion specifier: {ch}");
 
-        _ => {
-            eprintln!("unknown conversion specifier: {}", conv.spec);
             format_arg_string(conv, arg)
         }
-    }
+    };
+
+    Ok(vec)
 }
 
-fn do_printf(format: &str, args: &[String]) -> io::Result<()> {
-    let mut arg_pos = 0;
-    let mut output = String::with_capacity(format.len() * 2);
-    let blank = String::new();
+fn do_printf<'a>(
+    format: &[u8],
+    mut arguments: impl Iterator<Item = &'a [u8]>,
+) -> Result<(), Box<dyn Error>> {
+    let token_vec = tokenize_format_str(format)?;
 
-    let tokenlist = tokenize_format_str(format);
-    for token in tokenlist {
+    let mut output = Vec::<u8>::with_capacity(format.len() * 2_usize);
+
+    for token in token_vec {
         match token {
-            Token::Literal(s) => {
-                output.push_str(&s);
-            }
+            Token::Conversion(co) => {
+                let arg_str = arguments.next().unwrap_or(&[]);
 
-            Token::Conversion(c) => {
-                let arg_str = {
-                    if arg_pos >= args.len() {
-                        &blank
-                    } else {
-                        &args[arg_pos]
-                    }
-                };
-                arg_pos += 1;
+                let format_arg_result = format_arg(&co, arg_str)?;
 
-                output.push_str(format_arg(&c, arg_str).as_str());
+                output.extend_from_slice(format_arg_result.as_slice());
+            }
+            Token::IgnoreRestOfFormat => {
+                break;
+            }
+            Token::Literal(vec) => {
+                output.extend_from_slice(vec.as_slice());
             }
         }
     }
 
-    io::stdout().write_all(output.as_bytes())
+    io::stdout().write_all(output.as_slice())?;
+
+    Ok(())
 }
 
-fn main() -> Result<(), Box<dyn std::error::Error>> {
+fn start() -> Result<(), Box<dyn Error>> {
     setlocale(LocaleCategory::LcAll, "");
     textdomain(env!("PROJECT_NAME"))?;
     bind_textdomain_codeset(env!("PROJECT_NAME"), "UTF-8")?;
 
-    let args: Vec<String> = std::env::args().collect();
+    let args = std::env::args_os().collect::<Vec<_>>();
 
-    if args.len() < 2 {
-        return Err(gettext("printf: not enough arguments").into());
-    }
+    match &args.as_slice() {
+        &&[ref _program_file_name, ref format, ref arguments @ ..] => {
+            let arguments_iterator = arguments.iter().map(|os| os.as_bytes());
 
-    do_printf(&args[1], &args[2..])?;
+            do_printf(format.as_bytes(), arguments_iterator)?;
+        }
+        _ => {
+            return Err(Box::from(gettext("not enough arguments")));
+        }
+    }
 
     Ok(())
 }
+
+fn main() -> ExitCode {
```

Comment:
Let's not start a new style of `main` function.  This should be uniform with the other programs.

We can consider a `main` cleanup across multiple programs, assuming that makes sense, and not create a one-off style for this one program.

Reply:
Done. As you've probably seen, the issue is that when the a function returns a `Result`, Rust autogenerates some code that prints "Error: " and then the error, if an `Err` is returned.